pkg/controller/transaction/controller.go

```diff
@@ -0,0 +1,246 @@
+	c := controller.NewController("transaction")
+
+	c.Watch(&Watcher{
+		transactions: transactions,
+	})
+
+	c.Watch(&ConfigurationWatcher{
+		transactions:   transactions,
+		configurations: configurations,
+	})
+
+	c.Reconcile(&Reconciler{
+		transactions:   transactions,
+		topo:           topo,
+		configurations: configurations,
+		pluginRegistry: pluginRegistry,
+	})
+
+	return c
+}
+
+// Reconciler reconciles transactions
+type Reconciler struct {
+	topo           topo.Store
+	transactions   transaction.Store
+	configurations configuration.Store
+	pluginRegistry *pluginregistry.PluginRegistry
+}
+
+// Reconcile reconciles transactions
+func (r *Reconciler) Reconcile(id controller.ID) (controller.Result, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTimeout)
+	defer cancel()
+
+	transactionID := id.Value.(configapi.TransactionID)
+	transaction, err := r.transactions.Get(ctx, transactionID)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			log.Warnf("Failed to reconcile transaction %s, %s", transactionID, err)
+			return controller.Result{}, err
+		}
+		log.Debugf("Transaction %s not found", transactionID)
+		return controller.Result{}, nil
+	}
+
+	log.Infof("Reconciling transaction %v", transaction)
+	if ok, err := r.reconcileTransaction(ctx, transaction); err != nil {
+		log.Warnf("Failed to reconcile transaction: %s, %s", transactionID, err)
+		return controller.Result{}, err
+	} else if ok {
+		return controller.Result{}, nil
+	}
+
+	return controller.Result{}, nil
+}
+
+func (r *Reconciler) reconcileTransaction(ctx context.Context, transaction *configapi.Transaction) (bool, error) {
+	log.Debugf("Reconciling transaction %s in %s state", transaction.ID, transaction.Status.State)
+	switch transaction.Status.State {
+	case configapi.TransactionState_TRANSACTION_PENDING:
+		// If the transaction is Pending, begin validation if the prior transaction
+		// has already been applied. This simplifies concurrency control in the controller
+		// and guarantees transactions are applied to the configurations in sequential order.
+		return r.reconcileTransactionPending(ctx, transaction)
+		// if the transaction is in the Validating state, Validate the changes in a transaction and
+		// change the state to Applying if validation passed otherwise fail the transaction
+	case configapi.TransactionState_TRANSACTION_VALIDATING:
+		return r.reconcileTransactionValidating(ctx, transaction)
+		// If the transaction is in the Applying state, update the Configuration for each
+		// target and Complete the transaction.
+	case configapi.TransactionState_TRANSACTION_APPLYING:
+		return r.reconcileTransactionApplying(ctx, transaction)
+
+	}
+	return true, nil
+}
+
+func (r *Reconciler) reconcileTransactionPending(ctx context.Context, transaction *configapi.Transaction) (bool, error) {
+	if transaction.Index > 1 {
```

Comment:
Need to remove this condition. Checking the previous transaction is what we want, and this will prevent that from being done on the first transaction.